### PR TITLE
feat: Add `ALCF/examples/finetune_llama3/*`

### PR DESCRIPTION
**Pull request by saforem2**
Copilot Generated Summary:

This pull request includes several updates to the `ALCF/examples/finetune_llama3` directory, focusing on adding a new README file, configuration files, and a shell script for fine-tuning the Llama3 model. Additionally, there is a minor update in the `megatron/checkpointing.py` file to ensure directory creation when saving the learning rate state dictionary.

Documentation and setup:

* [`ALCF/examples/finetune_llama3/README.md`](diffhunk://#diff-7407bb198afeeccb703ba474fb8e1f547b289bd83bbc347f24d34c3adae5e6e1R1-R73): Added comprehensive instructions for setting up the environment, installing dependencies, downloading data, and converting Hugging Face checkpoints for fine-tuning Llama3.

Configuration files:

* [`ALCF/examples/finetune_llama3/ds_config.json`](diffhunk://#diff-718340babe725f11c9633d1b9df4f26e52222630a066492924ea3306e8f2b2d6R1-R18): Introduced a new DeepSpeed configuration file with parameters for batch size, optimizer, and bf16 support.
* [`ALCF/examples/finetune_llama3/ds_config_empty.json`](diffhunk://#diff-6ebfb6f9309a95114391ca58e789df20c49f7771da3b841edf746fcab9295dddR1-R18): Added an empty DeepSpeed configuration file template with similar parameters as `ds_config.json`.

Shell script:

* [`ALCF/examples/finetune_llama3/finetune_llama.sh`](diffhunk://#diff-95da00dc152f91d81ffec502e78f2a5314a3ad8d00258f11fe563e3f6bdbdca1R1-R220): Added a script for setting up the environment, configuring DeepSpeed, and running the fine-tuning process for Llama3. This script includes logic for converting Hugging Face models to Megatron-Deepspeed format and vice versa.

Minor update:

* [`megatron/checkpointing.py`](diffhunk://#diff-122925dfa160fba3c00803abba3577ef0d5aa5ab48989032a63d41c91f2a8002R589): Ensured the parent directory is created if it does not exist when saving the learning rate state dictionary.

### CODE REVIEW

File: tools/hf2megads_weight_converter.py [hunk 1]
```diff
@@ -57,85 +71,111 @@ def compute_partition_range(hidden_size, local_rank, tp_size):
...
+    )
 
-        for name, param in hf_auto_model.named_parameters():
-            hf_model[name] = param.clone()
-            log.info(name)
+    hf_model = {}
+    for name, submodule in hf_auto_model.named_children():
+        for pname, param in submodule.named_parameters():
+            logger.info(f'[{name}.{pname}] shape={param.shape}')
+            hf_model[f'{name}.{pname}'] = param.clone()
```
Comment (saforem2) [inline tools/hf2megads_weight_converter.py:135]: There was an issue being caused by the following block:

```python
for name, param in hf_auto_model.named_parameters():
```

which was failing to capture the `hf_auto_model.lm_head.weight`, which was preventing the checkpoint from being converted successfully.

Replacing this block with

```python
>>> for name, submodule in lmodel.named_children():
...     for pname, param in submodule.named_parameters():
...         named_submods[f'{name}.{pname}'] = param.clone()
```

fixes this issue, as shown explicitly below:

```python
>>> from transformers import AutoModelForCausalLM, LlamaConfig, AutoTokenizer, LlamaForCausalLM, AutoConfig
>>> lmodel = AutoModelForCausalLM.from_pretrained('meta-llama/Llama-3.2-1B')
>>> named_params = []
>>> named_params = {}
>>> for name, param in lmodel.named_parameters():
...     named_params[name] = param.clone()

>>> named_submods = {}
>>> for name, submodule in lmodel.named_children():
...     for pname, param in submodule.named_parameters():
...         named_submods[f'{name}.{pname}'] = param.clone()

>>> len(named_submods.keys())
147

>>> len(named_params.keys())
146

>>> list(named_submods.keys())[-3:]
['model.layers.15.post_attention_layernorm.weight',
 'model.norm.weight',
 'lm_head.weight']

>>> list(named_params.keys())[-3:]
['model.layers.15.input_layernorm.weight',
 'model.layers.15.post_attention_layernorm.weight',
 'model.norm.weight']
```

File: tools/hf2megads_weight_converter.py [hunk 2]
```diff
@@ -57,85 +71,111 @@ def compute_partition_range(hidden_size, local_rank, tp_size):
...
     def __init__(self, ds_model, hf_model, args, config):
-        tokenizer = get_tokenizer()
+        self.tokenizer = get_tokenizer()
+        if args.tokenizer_type == 'HFTokenizer':
+            self.hf_tokenizer = get_hf_tokenizer(args.tokenizer_model)
+            self.token_vocab = len(self.hf_tokenizer)
+        else:
+            self.hf_tokenizer = None
+            assert self.tokenizer is not None
+            self.token_vocab = self.tokenizer.vocab_size
```
Comment (saforem2) [inline tools/hf2megads_weight_converter.py:174]: Mismatch between `self.tokenizer.vocab_size` `hf_w.shape[0]` when using Llama3 tokenizers.

This discrepancy causes the following assertion:

```python
assert hf_w.shape[0] == self.padded_vocab_size
```

to fail since `hf_w.shape[0] = 128256` but `self.tokenizer.vocab_size = 128000`.

Explicitly:

```python
>>> type(self.hf_tokenizer)
<class 'transformers.tokenization_utils_fast.PreTrainedTokenizerFast'>
>>> type(self.tokenizer)
<class 'megatron.tokenizer.tokenizer._HFTokenizer'>
>>> self.tokenizer.vocab_size
128000
>>> len(self.hf_tokenizer)
128256
```

so, replacing `self.token_vocab = len(self.hf_tokenizer)` seems to resolve this issue.

**Comment by saforem2** [comment]
This pull request includes several updates to the `ALCF/examples/finetune_llama3` directory, focusing on enhancing the setup, configuration, and logging for fine-tuning Llama models. The key changes include adding a detailed README for setup instructions, updating configuration files, and improving logging and environment setup in the scripts.

### Documentation:
* [`ALCF/examples/finetune_llama3/README.md`](diffhunk://#diff-7407bb198afeeccb703ba474fb8e1f547b289bd83bbc347f24d34c3adae5e6e1R1-R73): Added detailed setup instructions for fine-tuning Llama3 from a Hugging Face checkpoint, including cloning the repository, setting up the environment, installing dependencies, downloading data, and converting checkpoints.

### Configuration:
* `ALCF/examples/finetune_llama3/ds_config.json` and `ALCF/examples/finetune_llama3/ds_config_empty.json`: Added configuration files for DeepSpeed, specifying training batch sizes, optimizer settings, and bf16 optimization. [[1]](diffhunk://#diff-718340babe725f11c9633d1b9df4f26e52222630a066492924ea3306e8f2b2d6R1-R18) [[2]](diffhunk://#diff-6ebfb6f9309a95114391ca58e789df20c49f7771da3b841edf746fcab9295dddR1-R18)

### Scripts:
* [`ALCF/examples/finetune_llama3/finetune_llama.sh`](diffhunk://#diff-95da00dc152f91d81ffec502e78f2a5314a3ad8d00258f11fe563e3f6bdbdca1R1-R232): Enhanced the script to include environment setup, dataset download, model configuration, and command execution for converting and fine-tuning models.

### Logging and Environment:
* [`finetune_llama.py`](diffhunk://#diff-40ba9864840fa8d5fefd32e95895823fa72cda19d7e9c62acc13128972eecb8cR5-L9): Integrated `ezpz` for environment setup and logging, added support for Weights & Biases (W&B) logging, and replaced `print_rank_0` with `logger.info`. [[1]](diffhunk://#diff-40ba9864840fa8d5fefd32e95895823fa72cda19d7e9c62acc13128972eecb8cR5-L9) [[2]](diffhunk://#diff-40ba9864840fa8d5fefd32e95895823fa72cda19d7e9c62acc13128972eecb8cR32-R63) [[3]](diffhunk://#diff-40ba9864840fa8d5fefd32e95895823fa72cda19d7e9c62acc13128972eecb8cL85-R131) [[4]](diffhunk://#diff-40ba9864840fa8d5fefd32e95895823fa72cda19d7e9c62acc13128972eecb8cL223-R269) [[5]](diffhunk://#diff-40ba9864840fa8d5fefd32e95895823fa72cda19d7e9c62acc13128972eecb8cL308-R354)
* [`megatron/data/prompt_dataset.py`](diffhunk://#diff-7214ada1001a097c778406d0984655eaae7c822edd458fea46a84ec5003da3a9R2-R15): Updated to use `ezpz` for logging instead of the standard logging module. [[1]](diffhunk://#diff-7214ada1001a097c778406d0984655eaae7c822edd458fea46a84ec5003da3a9R2-R15) [[2]](diffhunk://#diff-7214ada1001a097c778406d0984655eaae7c822edd458fea46a84ec5003da3a9L41-R55)

These changes collectively improve the usability, configuration, and logging of the fine-tuning process for Llama models.
